merkle/smt/writer.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smt
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/storage/tree"
+)
+
+// NodeBatchAccessor reads and writes batches of Merkle tree node hashes.
+type NodeBatchAccessor interface {
+	// Get returns the hashes of the given nodes, as the corresponding values of
+	// the passed-in map. The initial map values must be nil, and will remain nil
+	// for nodes that root empty subtrees.
+	Get(nodes map[tree.NodeID2][]byte) error
+	// Set applies the given node hash updates.
+	Set(upd []NodeUpdate) error
+}
+
+// Writer handles sharded writes to a sparse Merkle tree.
+type Writer struct {
+	treeID int64
+	hasher hashers.MapHasher
+	shards map[uint]shardWriter
+}
+
+// NewWriter creates a new Writer for the specified tree of the given height,
+// with 2 levels of sharding. One shard spans the the upper levels down to
+// split point, and all the other shards span the levels below.
+func NewWriter(treeID int64, hasher hashers.MapHasher, height, split uint) *Writer {
+	if split > height {
+		panic(fmt.Errorf("NewWriter: split(%d) > height(%d)", split, height))
+	}
+	shards := make(map[uint]shardWriter, 2)
+	w := &Writer{treeID: treeID, hasher: hasher, shards: shards}
+	shards[height] = shardWriter{w: w, depth: height, top: split}
+	shards[split] = shardWriter{w: w, depth: split, top: 0}
+	return w
+}
+
+// Split sorts and splits the given list of node hash updates into shards, i.e.
+// the subsets belonging to different subtrees. It calls the visit function for
+// each of the shards, and stops if the latter returns an error. The updates
+// must belong to the same tree level, one of the sharding split points. For
+// example, with a 2-level sharding it only makes sense to Split leaf updates.
+func (w *Writer) Split(upd []NodeUpdate, visit func([]NodeUpdate)) error {
```

Comment:
I would expect a function called `Split` to return splits. 
Should we separate responsibilities here for simplicity? 
Returning `[][]NodeUpdate` wouldn't waste memory since they're all indexed out of the same parent array. The caller could then do what it wants to with the split nodes.

Reply:
Done.